src/server/webkit.ts

```diff
@@ -88,143 +87,112 @@ export class WebKit extends BrowserTypeBase {
+
+  const server = new WebSocketServer(transport, logger, port, {
+    onBrowserResponse(seqNum: number, source: ws, message: ProtocolResponse) {
+      if (seqNum === -9999)
         return;
-      const { id, socket } = value;
-
-      if (socket.readyState === ws.CLOSING) {
-        if (pendingBrowserContextCreations.has(id)) {
-          transport.send({
-            id: ++SequenceNumberMixer._lastSequenceNumber,
-            method: 'Playwright.deleteContext',
-            params: { browserContextId: message.result.browserContextId }
-          });
-        }
+
+      if (source.readyState === ws.CLOSING) {
+        if (pendingBrowserContextCreations.has(seqNum))
+          server.sendMessageToBrowserOneWay('Playwright.deleteContext', { browserContextId: message.result.browserContextId });
         return;
       }
 
-      if (pendingBrowserContextCreations.has(message.id)) {
+      if (pendingBrowserContextCreations.has(seqNum)) {
         // Browser.createContext response -> establish context attribution.
-        browserContextIds.set(message.result.browserContextId, socket);
-        pendingBrowserContextCreations.delete(message.id);
+        browserContextIds.set(message.result.browserContextId, source);
+        pendingBrowserContextCreations.delete(seqNum);
       }
 
-      const deletedContextId = pendingBrowserContextDeletions.get(message.id);
+      const deletedContextId = pendingBrowserContextDeletions.get(seqNum);
       if (deletedContextId) {
         // Browser.deleteContext response -> remove context attribution.
         browserContextIds.delete(deletedContextId);
-        pendingBrowserContextDeletions.delete(message.id);
+        pendingBrowserContextDeletions.delete(seqNum);
       }
 
-      message.id = id;
-      socket.send(JSON.stringify(message));
+      source.send(JSON.stringify(message));
       return;
-    }
-
-    // Process notification response.
-    const { method, params, pageProxyId } = message;
-    if (pageProxyId) {
-      const socket = pageProxyIds.get(pageProxyId);
-      if (!socket || socket.readyState === ws.CLOSING) {
-        // Drop unattributed messages on the floor.
+    },
+
+    onBrowserNotification(message: ProtocolResponse) {
+      // Process notification response.
+      const { method, params, pageProxyId } = message;
+      if (pageProxyId) {
+        const data = pageProxyIds.get(pageProxyId);
+        if (!data || data.socket.readyState === ws.CLOSING) {
+          // Drop unattributed messages on the floor.
+          return;
+        }
+        data.socket.send(JSON.stringify(message));
         return;
       }
-      socket.send(JSON.stringify(message));
-      return;
-    }
-    if (method === 'Playwright.pageProxyCreated') {
-      const socket = browserContextIds.get(params.pageProxyInfo.browserContextId);
-      if (!socket || socket.readyState === ws.CLOSING) {
-        // Drop unattributed messages on the floor.
+      if (method === 'Playwright.pageProxyCreated') {
```

Comment:
We have seven 'Playwright.*' events and only handle 3 here :(

Reply:
Yes, not sure what you mean by the emoji. Are you saying there is a bug?

Reply:
Sorted out offline. Blocked by a WebKit patch.